module/zcommon/zfeature_common.c

```diff
@@ -559,6 +562,15 @@ zpool_feature_init(void)
 	    "com.datto:resilver_defer", "resilver_defer",
 	    "Support for deferring new resilvers when one is already running.",
 	    ZFEATURE_FLAG_READONLY_COMPAT, ZFEATURE_TYPE_BOOLEAN, NULL);
+
+	static const spa_feature_t zstd_deps[] = {
+		SPA_FEATURE_EXTENSIBLE_DATASET,
+		SPA_FEATURE_NONE
+	};
+	zfeature_register(SPA_FEATURE_ZSTD_COMPRESS,
+	    "org.freebsd:zstd_compress", "zstd_compress",
+	    "zstd compression algorithm support.",
+	    ZFEATURE_FLAG_PER_DATASET, ZFEATURE_TYPE_BOOLEAN, zstd_deps);
```

Comment:
We should set the flags to ZFEATURE_FLAG_READONLY_COMPAT | ZFEATURE_FLAG_PER_DATASET

So that it is still possible to import a pool with zstd on a system that does not support zstd. Any datasets not using zstd will still be usable.

Reply:
The problem with this approach is that we can not 100% guarantee it to be trouble free, but are expected to provide support.

Not allowing it manages expectations accordingly.

Reply:
The feature is definitely _not_read-only compatible. Is the `ZFEATURE_FLAG_READONLY_COMPAT | ZFEATURE_FLAG_PER_DATASET ` combination handled differently than those two behaviors individually? Otherwise, setting `ZFEATURE_FLAG_READONLY_COMPAT` seems wrong. An old implementation could load a pool with zstd compressed datasets and read garbage.

Reply:
@rlaager To add:
"reading garbage" can lead to all sorts of unexpected behavior...

Reply:
Maybe we need a new flag then. The idea being: I should be able to import a pool that has some zstd datasets. I will not be able to read or mount those datasets, but will be able to use the rest of the pool. Since zstd is a FLAG_PER_DATASET feature.

Maybe I am overreaching, but since zstd doesn't change anything outside of datasets that have this per-dataset feature referenced, it should be doable. It may not be worth the effort though.

Reply:
@allanjude 
Why a flag? I think it might be more interesting just mounting every dataset that has no flag issues (or an argument to do so) 

Simply put: It zfs (zstd or not) should be able to exclude incompatible datasets while importing pools anyway, regardless of zstd... If thats currently not supported maybe thats an additional PR someone should be writhing...